src/staging/validation.py

```diff
@@ -79,8 +85,26 @@ def validate_post_col(
             f"Invalid pattern postcodes found: {invalid_pattern_postcodes.to_list()}"
         )
 
+    # Remove the invalid pattern postcodes from the masterlist test
+    val_df = df.loc[~df.index.isin(invalid_pattern_postcodes.index.to_list())]
+
+    # Clean and harmonise the "601" postcodes to match the masterlist
+    check_real_df = clean_postcodes(val_df, "601")
+
+    # Only validate not null postcodes
+    check_real_df = check_real_df.loc[~check_real_df["601"].isnull()]
+
     # Create a list of postcodes not found in masterlist
-    unreal_postcodes = check_pcs_real(df, postcode_masterlist, config)
+    unreal_postcodes = check_pcs_real(df, check_real_df, postcode_masterlist, config)
+
+    # Create empty column for harmonised postcodes
+    df["postcodes_harmonised"] = np.nan
+
+    # Return harmonised postcodes for all rows of postcodes
+    harmonised_df = clean_postcodes(val_df, "postcodes_harmonised")
+    df.loc[harmonised_df.index, "postcodes_harmonised"] = harmonised_df[
+        "postcodes_harmonised"
+    ]
```

Comment:
```suggestion
    df.loc[harmonised_df.index, "postcodes_harmonised"] = clean_postcodes(val_df, "postcodes_harmonised")["postcodes_harmonised"]
```
Not sure if this suggestion is valid - could also change the output of `clean_postcodes` to be a column rather than a df.

Reply:
I run in to all sorts of indexing issues with the suggested method since the 2 don't match in size/shape. The current method does it in 2 steps but works fine

Reply:
Yeh that makes sense - keep it as is then, probably slightly more readable as well at present.